File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-1.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        m=len(matrix)
        n=len(matrix[0])
        visited=[[0]*n for i in range(m)]
        dp=[[0]*n for i in range(m)]

        def backtrack(i,j):
            if dp[i][j]:
                return dp[i][j]
            temp=matrix[i][j]
            matrix[i][j]=-1
            dp[i][j]=1

     

            if i+1<m and  matrix[i+1][j]!=-1 and matrix[i+1][j]>temp:
                dp[i][j]=max(dp[i][j],1+backtrack(i+1,j))
            if j+1<n and  matrix[i][j+1]!=-1 and matrix[i][j+1]>temp:
                dp[i][j]=max(dp[i][j],1+backtrack(i,j+1))
            if i-1>=0 and  matrix[i-1][j]!=-1 and matrix[i-1][j]>temp:
                dp[i][j]=max(dp[i][j],1+backtrack(i-1,j))
            if j-1>=0 and  matrix[i][j-1]!=-1 and matrix[i][j-1]>temp:
                dp[i][j]=max(dp[i][j],1+backtrack(i,j-1))
            matrix[i][j]=temp
            return dp[i][j]
            
        res=1
        for i in range(m):
            for j in range(n):
                res=max(res,backtrack(i,j))
        return res






        

File: Data Structures & Algorithms/count-number-of-islands/submission-1.java
class Solution {
    public void dfs(int i,int j,int m,int n,char[][] grid,int [][] visited){
            visited[i][j]=1;
            
            if(i+1<m && visited[i+1][j]!=1 && grid[i+1][j]=='1'){
                dfs(i+1,j,m,n,grid,visited);
            }
            if(j+1<n && visited[i][j+1]!=1 && grid[i][j+1]=='1'){
                dfs(i,j+1,m,n,grid,visited);
            }
            if(i-1>=0 && visited[i-1][j]!=1 && grid[i-1][j]=='1'){
                dfs(i-1,j,m,n,grid,visited);
            }
            if(j-1>=0 && visited[i][j-1]!=1 && grid[i][j-1]=='1'){
                dfs(i,j-1,m,n,grid,visited);
            }
            
        }
    public int numIslands(char[][] grid) {
        int m=grid.length;
        int n=grid[0].length;

        int visited[][]=new int[grid.length][grid[0].length];
        int res=0;

        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(grid[i][j]=='1' && visited[i][j]==0){
                    this.dfs(i,j,m,n,grid,visited);
                    res+=1;
                }

            }
        }

        

        return res;
        
    }
}
